tests/webap/test_webap.py

```diff
@@ -0,0 +1,157 @@
+
+        tabcontentdiv = browser.find_element( By.XPATH, "html/body/div/div/div/div/div/div/div[2]" )
+        assert tabcontentdiv.text[:15] == 'Exposures from '
+        explist = browser.find_element( By.ID, "exposure_list_table" )
+        rows = explist.find_elements( By.TAG_NAME, "tr" )
+        assert len(rows) == 2   # Just the header row
+
+        cols = rows[1].find_elements( By.XPATH, "./*" )
+        assert cols[0].text == 'c4d_230702_080904_ori.fits.fz'
+        assert cols[2].text == 'ELAIS-E1'
+        assert cols[5].text == '1'    # n_images
+        assert cols[6].text == '172'  # detections
+        assert cols[7].text == '6'    # sources
+
+        # Try to click on the exposure name, make sure we get the exposure details
+        expnamelink = cols[0].find_element( By.TAG_NAME, 'a' )
+        expnamelink.click()
+        WebDriverWait( browser, timeout=10 ).until(
+            lambda d: d.find_element( By.XPATH, "//h2[contains(.,'Exposure c4d_230702_080904_ori.fits.fz')]" ) )
+
+        # OMG I nest a lot of divs
+        tabcontentdiv = browser.find_element( By.XPATH, "html/body/div/div/div/div/div/div/div[2]" )
+        imagesdiv = tabcontentdiv.find_element( By.XPATH, "./div/div/div/div[2]/div" )
+        assert re.search( r"^Exposure has 1 images and 1 completed subtractions.*"
+                          r"6 out of 172 detections pass preliminary cuts",
+                          imagesdiv.text, re.DOTALL ) is not None
+
+
+        imagestab = imagesdiv.find_element( By.TAG_NAME, 'table' )
+        rows = imagestab.find_elements( By.TAG_NAME, 'tr' )
+        assert len(rows) == 2
+        cols = rows[1].find_elements( By.XPATH, "./*" )
+        assert re.search( r'^c4d_20230702_080904_S3_r_Sci', cols[1].text ) is not None
+
+        # Find the sources tab and click on that
+        tabbuttonsdiv = tabcontentdiv.find_element( By.XPATH, "./div/div/div/div[1]" )
+        sourcestab = tabbuttonsdiv.find_element( By.XPATH, "//.[.='Sources']" )
+        sourcestab.click()
+        # Give it half a second to go at least get to the "loading" screen; that's
+        #  all javascript with no server communcation, so should be fast.
+        time.sleep( 0.5 )
+        WebDriverWait( browser, timeout=10 ).until(
+            lambda d: d.find_element( By.XPATH, "//p[contains(.,'Sources for all successfully completed chips')]" ) )
+
+        # Now imagesdiv should have information about the sources
+        tabcontentdiv = browser.find_element( By.XPATH, "html/body/div/div/div/div/div/div/div[2]" )
+        imagesdiv = tabcontentdiv.find_element( By.XPATH, "./div/div/div/div[2]/div" )
+
+        sourcestab = imagesdiv.find_element( By.TAG_NAME, 'table' )
+        rows = sourcestab.find_elements( By.TAG_NAME, 'tr' )
+        assert len(rows) == 7
+        # check stuff about the rows?
+
+        # There is probably more we should be testing here.  Definitely.
```

Comment:
We can robustify these tests down the line I think, but these look like a good starting point.

Reply:
Lots of stuff is going to be added to the webap, so tests will be added then.